regionurlprovider.go

```diff
@@ -0,0 +1,159 @@
+package lksdk
+
+import (
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+	"sync"
+	"time"
+
+	"google.golang.org/protobuf/encoding/protojson"
+
+	"github.com/livekit/protocol/livekit"
+)
+
+const (
+	regionHostnameProviderSettingsCacheTime = 3 * time.Second
+)
+
+type regionURLProvider struct {
+	hostnameSettingsCache map[string]*hostnameSettingsCacheItem // hostname -> regionSettings
+
+	mutex      sync.RWMutex
+	httpClient *http.Client
+}
+
+type hostnameSettingsCacheItem struct {
+	regionSettings    *livekit.RegionSettings
+	updatedAt         time.Time
+	regionURLAttempts map[string]int
+}
+
+func newRegionURLProvider() *regionURLProvider {
+	return &regionURLProvider{
+		hostnameSettingsCache: make(map[string]*hostnameSettingsCacheItem),
```

Comment:
any specific reason to support multiple cloud host names? Guessing this is going to be used for a project.

Reply:
no specific reason. I just didn't know how people were going to be using this SDK and wanted to keep it flexible.